lib/tapioca/gem/listeners/source_location.rb

```diff
@@ -41,25 +55,31 @@ def on_method(event)
         def add_source_location_comment(node, file, line)
           return unless file && line
 
-          gem = @pipeline.gem
           path = Pathname.new(file)
           return unless File.exist?(path)
 
           # On native extensions, the source location may point to a shared object (.so, .bundle) file, which we cannot
           # use for jump to definition. Only add source comments on Ruby files
           return unless path.extname == ".rb"
 
-          path = if path.realpath.to_s.start_with?(gem.full_gem_path)
-            "#{gem.name}-#{gem.version}/#{path.realpath.relative_path_from(gem.full_gem_path)}"
+          gem = @pipeline.gem
+          realpath = path.realpath.to_s
+
+          # If the definition source location is not from the current gem, look for it in dependencies
+          gem = @gem_specs[realpath] unless realpath.start_with?(gem.full_gem_path)
+          return if gem.nil?
+
+          path = if gem.default_gem?
+            "#{gem.name}//#{path.realpath.relative_path_from(RbConfig::CONFIG["rubylibdir"])}"
           else
-            path.sub("#{Bundler.bundle_path}/gems/", "").to_s
+            "#{gem.name}//#{path.realpath.relative_path_from(gem.full_gem_path)}"
           end
 
-          # Strip out the RUBY_ROOT prefix, which is different for each user
-          path = path.sub(RbConfig::CONFIG["rubylibdir"], "RUBY_ROOT")
-
+          uri = URI("source://#{path}##{line}")
```

Comment:
Can we instead use `URI.build` and pass in all the components explicitly? I feel like we are still treating URI building as string building which isn't the best thing to do.

Reply:
You can see what I mean [here](https://github.com/Shopify/ruby-lsp/pull/237).

Reply:
Right, I see what you are doing there, and honestly I don't want to delay this PR longer, but I also feel like it would be so much more efficient and less code if we did almost the exact opposite of what we are doing here in the client.

Basically, the client can:
1. Using `Gem::Specification.default_stubs` and `.stubs`, build a hash with keys as `[gem_name, gem_version]` and values as a hash from relative file names to full file names.
2. When we are asked handle a source: URI, we parse it and grab the `gem_name`, which is guaranteed to always exist.
3. We then parse the `gem_version`, if it doesn't exist, then we parse it from the file name. We expect this to always end up with a gem version number, if not, we bail and show an error or something.
4. Now that we have a `[gem_name, gem_version]` we look it up in our table. If the gem name and version is also loaded, that gives us another hash table. If the gem name/version is not found, we fail gracefully.
5. Now we are left with a hashtable of relative file names to full file names, and we use the `path` of the URI to look up the full file path.
6. We open that file on the `line_number` from the URI.

This feels much easier to implement and way more robust than looking up files in various magical folders, and trying to guess the version of the gem, etc. WDYT?

Reply:
I agree with this approach. But if we look up in `default_stubs + stubs`, would we ever need to try to figure out the version from the file name?

Also, given this strategy, do you still wish to include the versions in the comments?

Reply:
I think there is still value and that is in cases where the RBI file has gone stale. Many projects are behind on updating this gem RBI files and we should be resilient to that. So including the version (when it can't be inferred from the filename) is a good thing for graceful handling of version mismatches. Basically, a gem's installed version could be drastically different from the RBI gem version and file paths/line numbers could have changed a lot in between those versions as well.

Specifically, if we change my suggestion above to be have an extra layer of lookup, with initial layer begin gem name and second layer being gem version and last layer being relative file path, then we can fail in this way:

1. If the lookup on gem name fails then LSP client can warn saying that the gem dependency has been removed and prompt to update RBI files.
2. If the gem version lookup fails, then we can prompt saying the gem version is different and file/line might have changed and ask if the user still wants to continue.
3. If the relative file path has changed, communicate that tot hw user and maybe again prompt for RBI generation again.